client/src/common/client.ts

```diff
@@ -317,6 +317,15 @@ interface _Middleware {
 	window?: WindowMiddleware;
 }
 
+// A general middleware is applied to both requests and notifications
+interface GeneralMiddleware {
+	provideRequestOrNotification?<R>(
```

Comment:
I would prefer having a separate hook for notifications and requests. I would also like to suggest to not name it provide since they don't provide anything. Better would be something like:

```
  sendRequest(....)
  sendNotification(...)
```

Reply:
Separated the two and renamed them.